### PR TITLE
Rewrite the cryptree serialization

### PR DESCRIPTION
**Pull request by ianopolous**
This does a big clean up of the cryptree serialization and includes a few breaking changes. 

The major change is the bulk padding and encryption of child links on directories. This is to obscure the exact number of children present, which makes reconstructing directory tree topologies much more difficult. 

The other big change is now the entry points in the static data are versioned, padded and bulk encrypted. This obscures further the number of friends a user has, and also paves the way for plausibly deniable dual log in.

### CODE REVIEW

File: src/peergos/shared/user/fs/DirAccess.java [hunk 1]
```diff
@@ -134,304 +123,211 @@ public static DirAccess fromCbor(CborObject cbor, Multihash hash) {
...
+            try {
+                Location target = e.location;
+                boolean keep = true;
+                if (Arrays.equals(target.getMapKey(), childRetrievedPointer.capability.location.getMapKey()))
+                    if (Arrays.equals(target.writer.serialize(), childRetrievedPointer.capability.location.writer.serialize()))
+                        if (Arrays.equals(target.owner.serialize(), childRetrievedPointer.capability.location.owner.serialize()))
+                            keep = false;
+                return keep;
+            } catch (TweetNaCl.InvalidCipherTextException ex) {
+                ex.printStackTrace();
```
Comment (kevodwyer) [inline src/peergos/shared/user/fs/DirAccess.java]: ex required?

File: src/peergos/shared/crypto/PaddedCipherText.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    public static PaddedCipherText fromCbor(Cborable cbor) {
+        return new PaddedCipherText(CipherText.fromCbor(cbor));
+    }
+
+    private static byte[] pad(byte[] input, int blockSize) {
+        int nBlocks = (input.length + blockSize - 1) / blockSize;
+        return Arrays.copyOfRange(input, 0, nBlocks * blockSize);
+    }
+
+    public static <T extends Cborable> PaddedCipherText build(SymmetricKey from, T secret, int paddingBlockSize) {
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/shared/crypto/PaddedCipherText.java]: done

File: src/peergos/server/tests/MultiUserTests.java [hunk 3]
```diff
@@ -246,20 +246,18 @@ public void cleanRenamedFiles() throws Exception {
...
-        Optional<FileTreeNode> unsharedView2 = userToUnshareWith.getByPath(friendsNewPathToFile).get();
+        Optional<FileWrapper> unsharedView2 = userToUnshareWith.getByPath(friendsNewPathToFile).get();
         CryptreeNode fileAccess = network.getMetadata(priorPointer.getLocation()).get().get();
         try {
             // Try decrypting the new metadata with the old key
-            byte[] properties = ((CborObject.CborByteArray) ((CborObject.CborList) fileAccess.toCbor()).value.get(1)).value;
-            byte[] nonce = Arrays.copyOfRange(properties, 0, TweetNaCl.SECRETBOX_NONCE_BYTES);
-            byte[] cipher = Arrays.copyOfRange(properties, TweetNaCl.SECRETBOX_NONCE_BYTES, properties.length);
-            FileProperties props =  FileProperties.fromCbor(CborObject.fromByteArray(priorMetaKey.decrypt(cipher, nonce)));
+            PaddedCipherText properties = PaddedCipherText.fromCbor(((CborObject.CborList) fileAccess.toCbor()).value.get(4));
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/server/tests/MultiUserTests.java]: done

File: src/peergos/shared/user/fs/DirAccess.java (hunk 1 above)
Comment (ianopolous) [reply src/peergos/shared/user/fs/DirAccess.java]: I've cleaned up that section

File: src/peergos/shared/user/fs/RetrievedCapability.java [hunk 4]
```diff
@@ -1,57 +1,63 @@
 package peergos.shared.user.fs;
 
-import peergos.shared.cbor.CborObject;
```
Comment (cboddy) [inline src/peergos/shared/user/fs/RetrievedCapability.java]: did you mena to update the permissions of `` src/peergos/shared/user/fs/RetrievedCapability.java``?

File: src/peergos/server/tests/MultiUserTests.java [hunk 5]
```diff
@@ -29,7 +29,7 @@
 @RunWith(Parameterized.class)
 public class MultiUserTests {
 
-    private static Args args = UserTests.buildArgs().with("useIPFS", "true");
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/server/tests/MultiUserTests.java]: These test don't need to use IPFS. Using ipfs just makes them slower, platform specific, and fail randomly on teamcity.

File: src/peergos/server/Playground.java [hunk 6]
```diff
@@ -66,7 +66,7 @@ private static void experiment(String username,
                 continue;
             SecretSigningKey signingKey = context.getUserRoot().get().getEntryWriterKey().get();
             SigningPrivateKeyAndPublicHash writer = new SigningPrivateKeyAndPublicHash(ownedKey, signingKey);
-            existing.props.migrateToChamp(context.signer.publicKeyHash, writer, existing.hash, network, res -> {}).get();
+            // Do something risky
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/server/Playground.java]: I used this to test out the migration from btree to champ on the demo server. This had the potential to lose all my data, hence testing it out first this way.

File: src/peergos/shared/user/UserStaticData.java [hunk 7]
```diff
@@ -2,92 +2,76 @@
...
+        }
 
-    public static UserStaticData fromCbor(Cborable cbor, SymmetricKey rootKey) {
-        if (! (cbor instanceof CborObject.CborByteArray))
-            throw new IllegalStateException("UserStaticData cbor must be a byte[]! " + cbor);
-        return deserialize(((CborObject.CborByteArray) cbor).value, rootKey);
+        public static EntryPoints fromCbor(Cborable cbor) {
+            if (! (cbor instanceof CborObject.CborMap))
+                throw new IllegalStateException("Incorrect cbor type for EntryPoints: " + cbor);
+            CborObject.CborLong version = (CborObject.CborLong)((CborObject.CborMap) cbor).values.get(new CborObject.CborString("v"));
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/shared/user/UserStaticData.java]: I agree. In another pr though.

File: src/peergos/shared/crypto/SymmetricLink.java [hunk 8]
```diff
@@ -2,46 +2,29 @@
...
 public class SymmetricLink implements Cborable
 {
-    private final byte[] nonce, link;
-
-    public SymmetricLink(byte[] link)
-    {
-        this.nonce = Arrays.copyOfRange(link, 0, TweetNaClKey.NONCE_BYTES);
-        this.link = Arrays.copyOfRange(link, TweetNaClKey.NONCE_BYTES, link.length);
-    }
+    private final CipherText cipherText;
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/shared/crypto/SymmetricLink.java]: Good catch. I've added javadoc in a later pr.

File: src/peergos/shared/user/UserStaticData.java [hunk 9]
```diff
@@ -2,92 +2,76 @@
...
-import peergos.shared.user.fs.Capability;
-import peergos.shared.util.*;
 
-import java.io.*;
 import java.util.*;
+import java.util.stream.*;
 
 public class UserStaticData implements Cborable {
 	private static final Logger LOG = Logger.getGlobal();
+	private static final int PADDING = 8192;
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply src/peergos/shared/user/UserStaticData.java]: Better name. Done in a later pr.

File: src/peergos/shared/user/fs/RetrievedCapability.java (hunk 4 above)
Comment (ianopolous) [reply src/peergos/shared/user/fs/RetrievedCapability.java]: Not sure what you mean?

**Pull request merged**